lifecycle.go

```diff
@@ -215,33 +215,39 @@ var defaultReadinessHook = func() ContainerLifecycleHooks {
 				b.MaxElapsedTime = 1 * time.Second
 				b.MaxInterval = 5 * time.Second
 
-				err := backoff.Retry(func() error {
-					jsonRaw, err := dockerContainer.inspectRawContainer(ctx)
-					if err != nil {
-						return err
-					}
+				err := backoff.RetryNotify(
+					func() error {
+						jsonRaw, err := dockerContainer.inspectRawContainer(ctx)
+						if err != nil {
+							return err
+						}
 
-					exposedAndMappedPorts := jsonRaw.NetworkSettings.Ports
-
-					for _, exposedPort := range dockerContainer.exposedPorts {
-						portMap := nat.Port(exposedPort)
-						// having entries in exposedAndMappedPorts, where the key is the exposed port,
-						// and the value is the mapped port, means that the port has been already mapped.
-						if _, ok := exposedAndMappedPorts[portMap]; !ok {
-							// check if the port is mapped with the protocol (default is TCP)
-							if !strings.Contains(exposedPort, "/") {
-								portMap = nat.Port(fmt.Sprintf("%s/tcp", exposedPort))
-								if _, ok := exposedAndMappedPorts[portMap]; !ok {
+						exposedAndMappedPorts := jsonRaw.NetworkSettings.Ports
+
+						for _, exposedPort := range dockerContainer.exposedPorts {
+							portMap := nat.Port(exposedPort)
+							// having entries in exposedAndMappedPorts, where the key is the exposed port,
+							// and the value is the mapped port, means that the port has been already mapped.
+							if _, ok := exposedAndMappedPorts[portMap]; !ok {
+								// check if the port is mapped with the protocol (default is TCP)
+								if !strings.Contains(exposedPort, "/") {
+									portMap = nat.Port(fmt.Sprintf("%s/tcp", exposedPort))
+									if _, ok := exposedAndMappedPorts[portMap]; !ok {
+										return fmt.Errorf("port %s is not mapped yet", exposedPort)
+									}
+								} else {
 									return fmt.Errorf("port %s is not mapped yet", exposedPort)
 								}
-							} else {
-								return fmt.Errorf("port %s is not mapped yet", exposedPort)
 							}
 						}
-					}
 
-					return nil
-				}, b)
+						return nil
+					},
+					b,
+					func(err error, duration time.Duration) {
+						dockerContainer.logger.Printf("Readiness hook: %v", err)
```

Comment:
Do we want to print a more consistent error here?
```suggestion
						dockerContainer.logger.Printf("All requested ports were not exposed: %v", err)
```

Reply:
There may be other errors too, from `inspectRawContainer(ctx)`. Not a big deal, can change, but that's why I kept it generic.

Reply:
Indeed, although the sole reason for calling the inspect would be checking the available exposed ports. Because the returned error would be added to the message, I think it's fine to combine both into the print.

If you mind, I'd prefer being explicit here

Reply:
No problem, updated!